include/dt-bindings/gpio/gpio.h

```diff
@@ -1,180 +1,429 @@
+ * The `GPIO_DIR_*` values derived from these flags may be more
+ * convenient.
+ *
  * @{
  */
-/** GPIO pin to trigger interrupt. */
-#define GPIO_INT		(1 << 1)
+/** @cond INTERNAL_HIDDEN */
+#define GPIO_IO_POS             5
+#define GPIO_IO_MASK            (0x0F << GPIO_IO_POS)
+/** @endcond */
 
-/** GPIO pin trigger on level low or falling edge. */
-#define GPIO_INT_ACTIVE_LOW	(0 << 2)
+/** Mask to isolate the `GPIO_IO_IN_ENABLED` and `GPIO_IO_OUT_ENABLED`
+ * flags.
+ */
+#define GPIO_DIR_MASK           (0x03 << GPIO_IO_POS)
 
-/** GPIO pin trigger on level high or rising edge. */
-#define GPIO_INT_ACTIVE_HIGH	(1 << 2)
+/** GPIO pin input is connected. */
+#define GPIO_IO_IN_ENABLED      (1 << GPIO_IO_POS)
 
-/** Enable GPIO pin debounce. */
-#define GPIO_INT_DEBOUNCE       (1 << 4)
+/** GPIO pin output is driven. */
+#define GPIO_IO_OUT_ENABLED     (1 << (GPIO_IO_POS + 1))
 
-/** Do Level trigger. */
-#define GPIO_INT_LEVEL		(0 << 5)
+/** GPIO pin output is driven. */
+#define GPIO_IO_INIT_ENABLED    (1 << (GPIO_IO_POS + 2))
 
-/** Do Edge trigger. */
-#define GPIO_INT_EDGE		(1 << 5)
+/** GPIO pin output is driven. */
+#define GPIO_IO_INIT_LOW        (0 << (GPIO_IO_POS + 3))
 
-/** Interrupt triggers on both rising and falling edge.
- *  Must be combined with GPIO_INT_EDGE.
+/** GPIO pin output is driven. */
+#define GPIO_IO_INIT_HIGH       (1 << (GPIO_IO_POS + 3))
+
+/** GPIO pin configured as input only. */
+#define GPIO_DIR_IN             GPIO_IO_IN_ENABLED
+
+/** GPIO pin configured as output with no change to output state. */
+#define GPIO_DIR_OUT            GPIO_IO_OUT_ENABLED
+
+/** GPIO pin configured as output low.
+ *
+ * Where configuration is not atomic the output value should be set
+ * before the pin is configured to drive the output.
+ */
+#define GPIO_DIR_OUT_LOW        (GPIO_IO_OUT_ENABLED	\
+				 | GPIO_IO_INIT_ENABLED	\
+				 | GPIO_IO_INIT_LOW)
+
+/** GPIO pin configured as output high.
+ *
+ * Where configuration is not atomic the output value should be set
+ * before the pin is configured to drive the output.
  */
-#define GPIO_INT_DOUBLE_EDGE	(1 << 6)
+#define GPIO_DIR_OUT_HIGH       (GPIO_IO_OUT_ENABLED	\
+				 | GPIO_IO_INIT_ENABLED	\
+				 | GPIO_IO_INIT_HIGH)
 /** @} */
 
 
 /**
- * @name GPIO polarity flags
- * The `GPIO_POL_*` flags are used with `gpio_pin_configure` or `gpio_port_configure`,
- * to specify the polarity of a GPIO pin.
+ * @name GPIO interrupt flags
+ *
+ * The `GPIO_INT_*` flags are used with `gpio_pin_configure` or
+ * `gpio_port_configure`, to specify how input GPIO pins will trigger
+ * interrupts.
+ *
  * @{
  */
 /** @cond INTERNAL_HIDDEN */
-#define GPIO_POL_POS		7
+#define GPIO_INT_POS            9
+#define GPIO_INT_MASK           (0x1F << GPIO_INT_POS)
 /** @endcond */
 
-/** GPIO pin polarity is normal. */
-#define GPIO_POL_NORMAL		(0 << GPIO_POL_POS)
+/** Mask to isolate the `GPIO_INT_LOW` and `GPIO_INT_HIGH` flags. */
+#define GPIO_INT_DIR_MASK       (0x03 << (GPIO_INT_POS + 2))
```

Comment:
`INT_DIR` looks a bit strange to me (is it "interrupt direction"?). But will this mask be actually useful, where you see the need to isolate this two flags?

Reply:
This would be used instead of `GPIO_INT_DOUBLE_EDGE` in conditions like this:
```
switch(flags & GPIO_INT_DIR_MASK) {
default: // invalid interrupt specification
case GPIO_INT_LOW: // detect low
case GPIO_INT_HIGH: // detect high
case GPIO_INT_LOW | GPIO_INT_HIGH: // detect both
}
```
`GPIO_INT_DOUBLE_EDGE` is currently equivalent to the last case, but it also allows some enhancements (e.g. its absence can indicate use of a low-precision level-based double edge detection approach on Nordic, useful when monitoring a large number of dry contacts).

Reply:
But in the above code you could do `(flags & (GPIO_INT_LOW | GPIO_INT_HIGH))`, what would be equally readable (at least to me).

Reply:
I don't understand.  All `flags & (GPIO_INT_LOW | GPIO_INT_HIGH)` tells you is that interrupts are supposed to be enabled.  It doesn't let know how which direction to configure the hardware.  (In the switch statement I didn't provide bodies for each case, but if they were there they'd all be different.)

Reply:
OK, I see; yes, you could do `flags & (GPIO_INT_LOW | GPIO_INT_HIGH)` instead of `flags & GPIO_INT_DIR_MASK` since they have the same effect.  I think the `flags & FOO_MASK` idiom is more clear, as it's obvious you're interested in the `FOO` field without requiring the code extracting it to know all the bits that might be present in it, or how they combine to define distinct values.